qiskit/transpiler/passes/scheduling/time_unit_conversion.py

```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
             return "SI"
 
         return "mixed"
+
+
+_DURATION_KIND_NAME = {
+    Duration.dt: "dt",
+    Duration.ns: "ns",
+    Duration.us: "us",
+    Duration.ms: "ms",
+    Duration.s: "s",
+}
+
+
+class _EvalDurationImpl(expr.ExprVisitor[float]):
+    """Evaluates the expression to a single float result.
+
+    If ``dt`` is provided or all durations are already in ``dt``, the result is in ``dt``.
+    Otherwise, the result will be in seconds, and all durations MUST be in wall-time (SI).
+    """
+
+    __slots__ = ("dt", "has_dt", "has_si")
+
+    def __init__(self, dt=None):
+        self.dt = dt if dt is not None else 1
+        self.has_dt = False
+        self.has_si = False
+
+    def in_cycles(self):
+        """Returns ``True`` if units are 'dt' after visit."""
+        return self.has_dt or self.dt != 1
+
+    def visit_value(self, node, /) -> float:
+        if isinstance(node.value, float):
+            return node.value
+        if isinstance(node.value, Duration.dt):
+            if self.has_si and self.dt == 1:
+                raise TranspilerError(
+                    "Fail to unify time units in delays. SI units "
+                    "and dt unit must not be mixed when dt is not supplied."
+                )
+            self.has_dt = True
+            return node.value[0]
+        if isinstance(node.value, Duration):
+            if self.has_dt and self.dt == 1:
+                raise TranspilerError(
+                    "Fail to unify time units in delays. SI units "
+                    "and dt unit must not be mixed when dt is not supplied."
+                )
+            self.has_si = True
+            if isinstance(node.value, Duration.s):
+                return node.value[0] / self.dt
+            from_unit = _DURATION_KIND_NAME.get(type(node.value))
+            return apply_prefix(node.value[0], from_unit) / self.dt
+        raise TranspilerError(f"invalid duration expression: {node}")
+
+    def visit_binary(self, node, /) -> float:
+        left = node.left.accept(self)
+        right = node.right.accept(self)
+        if node.op == expr.Binary.Op.ADD:
+            return left + right
+        if node.op == expr.Binary.Op.SUB:
+            return left - right
+        if node.op == expr.Binary.Op.MUL:
+            return left * right
+        if node.op == expr.Binary.Op.DIV:
+            return left / right
+        raise TranspilerError(f"invalid duration expression: {node}")
+
+    def visit_cast(self, node, /) -> float:
+        return node.operand.accept(self)
+
+    def visit_index(self, node, /) -> float:
+        return node.target.accept(self)
```

Comment:
Probably we ought to leave this undefined and/or let it error out - it can't currently appear in valid expressions of type `Duration`, and this implementation wouldn't be valid (doesn't examine the index) for most hypothetical working implementations.

Reply:
Done in 60e6b5cc12db9a89599f3cd3e16dc3f51291a5fb